src/System.Private.CoreLib/shared/System/Threading/Mutex.cs

```diff
@@ -0,0 +1,218 @@
+        // MUTEX_MODIFY_STATE rights (0x0001). The other interesting value
+        // in a Mutex's ACL is MUTEX_ALL_ACCESS (0x1F0001).
+        public void ReleaseMutex()
+        {
+#if CORECLR
+            if (!Win32Native.ReleaseMutex(safeWaitHandle))
+            {
+                throw new ApplicationException(SR.Arg_SynchronizationLockException);
+            }
+#else
+            // The field value is modifiable via the public <see cref="WaitHandle.SafeWaitHandle"/> property, save it locally
+            // to ensure that one instance is used in all places in this method
+            SafeWaitHandle waitHandle = _waitHandle;
+            if (waitHandle == null)
+            {
+                ThrowInvalidHandleException();
+            }
+
+            waitHandle.DangerousAddRef();
+            try
+            {
+                ReleaseMutexCore(waitHandle.DangerousGetHandle());
+            }
+            finally
+            {
+                waitHandle.DangerousRelease();
+            }
+#endif
+        }
+
+#if CORECLR || !PLATFORM_UNIX
+        private static void VerifyNameForCreate(string name)
+        {
+#if !PLATFORM_UNIX
+            if (name != null && (Interop.Kernel32.MAX_PATH < name.Length))
+            {
+                throw new ArgumentException(SR.Format(SR.Argument_WaitHandleNameTooLong, name, Interop.Kernel32.MAX_PATH), nameof(name));
+            }
+#endif
+        }
+
+        private void CreateMutexCore(bool initiallyOwned, string name, out bool createdNew)
+        {
+#if !PLATFORM_UNIX
+            Debug.Assert(name == null || name.Length <= Interop.Kernel32.MAX_PATH);
+#endif
+
+#if CORECLR
+            uint mutexFlags = initiallyOwned ? Win32Native.CREATE_MUTEX_INITIAL_OWNER : 0;
+            SafeWaitHandle mutexHandle = Win32Native.CreateMutexEx(null, name, mutexFlags, AccessRights);
+#else
+            uint mutexFlags = initiallyOwned ? (uint)Interop.Constants.CreateMutexInitialOwner : 0;
+            SafeWaitHandle mutexHandle = Interop.mincore.CreateMutexEx(IntPtr.Zero, name, mutexFlags, AccessRights);
+#endif
+            int errorCode = Marshal.GetLastWin32Error();
+
+            if (mutexHandle.IsInvalid)
+            {
+                mutexHandle.SetHandleAsInvalid();
+#if PLATFORM_UNIX
+                if (errorCode == Interop.Errors.ERROR_FILENAME_EXCED_RANGE)
+                    // On Unix, length validation is done by CoreCLR's PAL after converting to utf-8
+                    throw new ArgumentException(SR.Format(SR.Argument_WaitHandleNameTooLong, name, WaitHandleNameMax), nameof(name));
+#endif
+                if (errorCode == Interop.Errors.ERROR_INVALID_HANDLE)
+                    throw new WaitHandleCannotBeOpenedException(SR.Format(SR.Threading_WaitHandleCannotBeOpenedException_InvalidHandle, name));
+#if CORECLR
+                throw Win32Marshal.GetExceptionForWin32Error(errorCode, name);
+#else
+                throw ExceptionFromCreationError(errorCode, name);
```

Comment:
I think you should be able to delete `ExceptionFromCreationError` and just use `Win32Marshal.GetExceptionForWin32Error`. As far as I can tell, they should behave identically here.